File: src/cli/commands/add.ts
/**
 * Add command implementation
 * 
 * Adds a new host entry to ~/.hosts
 */

import { readHostsFile, writeHostsFile } from "../../core/file-io";
import { validateIP, validateHostname, validateNoDuplicates } from "../../domain/validators";
import { generateId } from "../../domain/id";
import type { Entry, Group, HostsFile } from "../../domain/types";

export type AddOptions = {
  group?: string;
  disabled?: boolean;
  comment?: string;
  hostsFile?: string;
};

/**
 * Execute the add command
 * 
 * @param ip - IP address (IPv4 or IPv6)
 * @param hostname - Primary hostname
 * @param aliases - Array of alias hostnames
 * @param options - Command options
 * @returns Exit code (0 = success, 1 = validation error, 2 = I/O error)
 */
export async function addCommand(
  ip: string,
  hostname: string,
  aliases: string[],
  options: AddOptions = {}
): Promise<number> {
  try {
    // Validate IP address
    const ipValidation = validateIP(ip);
    if (!ipValidation.valid) {
      console.error(`Error: ${ipValidation.error}`);
      return 1;
    }

    // Validate hostname
    const hostnameValidation = validateHostname(hostname);
    if (!hostnameValidation.valid) {
      console.error(`Error: ${hostnameValidation.error}`);
      return 1;
    }

    // Validate aliases
    for (const alias of aliases) {
      const aliasValidation = validateHostname(alias);
      if (!aliasValidation.valid) {
        console.error(`Error: Invalid alias "${alias}" - ${aliasValidation.error}`);
        return 1;
      }
    }

    // Read existing hosts file
    const hostsFilePath = options.hostsFile || "~/.hosts";
    const hostsFile = await readHostsFile(hostsFilePath);

    // Create new entry
    const newEntry: Entry = {
      id: generateId(),
      ip,
      hostname,
      aliases,
      enabled: !options.disabled,
      comment: options.comment,
    };

    // Check for duplicate hostnames
    const allEntries = collectAllEntries(hostsFile);
    const duplicateCheck = validateNoDuplicates([...allEntries, newEntry]);
    if (!duplicateCheck.valid) {
      console.error(`Error: ${duplicateCheck.error}`);
      return 1;
    }

    // Add entry to the appropriate location
    if (options.group) {
      // Add to specified group (create group path if needed)
      addEntryToGroup(hostsFile, newEntry, options.group);
    } else {
      // Add to root level - we need a synthetic root group or root entries array
      // Based on the design, we'll add a root group if it doesn't exist
      addEntryToRoot(hostsFile, newEntry);
    }

    // Write back to file
    await writeHostsFile(hostsFilePath, hostsFile);

    console.log(`✓ Added ${hostname} (${ip})`);
    return 0;

  } catch (err: any) {
    // Handle I/O errors
    if (err.code === "EACCES" || err.code === "ENOENT" || err.code === "EPERM") {
      console.error(`Error: ${err.message}`);
      return 2;
    }

    // Handle other errors
    console.error(`Error: ${err.message}`);
    return 2;
  }
}

/**
 * Collect all entries from all groups recursively
 */
function collectAllEntries(hostsFile: HostsFile): Entry[] {
  const entries: Entry[] = [];
  
  function collectFromGroup(group: Group) {
    entries.push(...group.entries);
    for (const subgroup of group.groups) {
      collectFromGroup(subgroup);
    }
  }
  
  for (const group of hostsFile.groups) {
    collectFromGroup(group);
  }
  
  return entries;
}

/**
 * Add entry to root level (ungrouped entries)
 */
function addEntryToRoot(hostsFile: HostsFile, entry: Entry) {
  // Look for a root/ungrouped group, or create one
  let rootGroup = hostsFile.groups.find(g => g.name === "");
  
  if (!rootGroup) {
    rootGroup = {
      name: "",
      entries: [],
      groups: [],
    };
    hostsFile.groups.push(rootGroup);
  }
  
  rootGroup.entries.push(entry);
}

/**
 * Add entry to a specific group path, creating groups as needed
 */
function addEntryToGroup(hostsFile: HostsFile, entry: Entry, groupPath: string) {
  const pathSegments = groupPath.split("/").filter(s => s.length > 0);
  
  let currentGroups = hostsFile.groups;
  let currentGroup: Group | undefined;
  
  for (const segment of pathSegments) {
    currentGroup = currentGroups.find(g => g.name === segment);
    
    if (!currentGroup) {
      currentGroup = {
        name: segment,
        entries: [],
        groups: [],
      };
      currentGroups.push(currentGroup);
    }
    
    currentGroups = currentGroup.groups;
  }
  
  // Add entry to the final group
  if (currentGroup) {
    currentGroup.entries.push(entry);
  }
}

File: go/internal/cmd/add.go
package cmd

import (
	"fmt"
	"os"

	"github.com/hungthai1401/hostie/go/internal/core/fileio"
	"github.com/hungthai1401/hostie/go/internal/domain"
	"github.com/spf13/cobra"
)

var (
	addDisabled bool
	addComment  string
	addGroup    string
)

func newAddCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "add <ip> <hostname> [aliases...]",
		Short: "Add a new hosts entry",
		Long:  "Add a new entry to the hosts file with optional aliases and group assignment",
		Args:  cobra.MinimumNArgs(2),
		RunE:  runAdd,
	}

	cmd.Flags().BoolVar(&addDisabled, "disabled", false, "add entry in disabled state")
	cmd.Flags().StringVarP(&addComment, "comment", "c", "", "comment for the entry")
	cmd.Flags().StringVarP(&addGroup, "group", "g", "", "group to add entry to")

	return cmd
}

func runAdd(cmd *cobra.Command, args []string) error {
	ip := args[0]
	hostname := args[1]
	aliases := args[2:]

	// Validate IP address
	if err := domain.ValidateIP(ip); err != nil {
		return fmt.Errorf("invalid IP address: %w", err)
	}

	// Validate hostname
	if err := domain.ValidateHostname(hostname); err != nil {
		return fmt.Errorf("invalid hostname: %w", err)
	}

	// Validate aliases
	if err := domain.ValidateAliases(aliases); err != nil {
		return err
	}

	// Read existing hosts file
	hostsFile, err := fileio.ReadHostsFile(hostsFilePath)
	if err != nil {
		// If file doesn't exist, create a new one
		if os.IsNotExist(err) {
			hostsFile = domain.HostsFile{
				Version: 1,
				Groups:  []domain.Group{},
			}
		} else {
			return fmt.Errorf("failed to read hosts file: %w", err)
		}
	}

	// Create new entry
	newEntry := domain.Entry{
		ID:       domain.NewID(),
		IP:       ip,
		Hostname: hostname,
		Aliases:  aliases,
		Enabled:  !addDisabled,
		Comment:  addComment,
	}

	// Check for duplicate hostnames
	allEntries := collectAllEntries(&hostsFile)
	allEntries = append(allEntries, newEntry)
	if err := domain.ValidateNoDuplicates(allEntries); err != nil {
		return err
	}

	// Add entry to the appropriate location
	if addGroup != "" {
		if err := addEntryToGroup(&hostsFile, newEntry, addGroup); err != nil {
			return err
		}
	} else {
		addEntryToRoot(&hostsFile, newEntry)
	}

	// Write back to file
	if err := fileio.WriteHostsFile(hostsFilePath, hostsFile); err != nil {
		return fmt.Errorf("failed to write hosts file: %w", err)
	}

	fmt.Printf("✓ Added %s (%s)\n", hostname, ip)
	return nil
}

// collectAllEntries recursively collects all entries from the hosts file
func collectAllEntries(hf *domain.HostsFile) []domain.Entry {
	var entries []domain.Entry
	for _, g := range hf.Groups {
		entries = append(entries, collectGroupEntries(&g)...)
	}
	return entries
}

// collectGroupEntries recursively collects entries from a group and its subgroups
func collectGroupEntries(g *domain.Group) []domain.Entry {
	entries := make([]domain.Entry, len(g.Entries))
	copy(entries, g.Entries)
	for _, subgroup := range g.Groups {
		entries = append(entries, collectGroupEntries(&subgroup)...)
	}
	return entries
}

// addEntryToRoot adds an entry to the root level (creates a default group if needed)
func addEntryToRoot(hf *domain.HostsFile, entry domain.Entry) {
	// If there are no groups, create a default one
	if len(hf.Groups) == 0 {
		hf.Groups = []domain.Group{
			{
				Name:        "default",
				Description: "Default group",
				Entries:     []domain.Entry{entry},
				Groups:      []domain.Group{},
			},
		}
		return
	}

	// Add to the first group (default behavior)
	hf.Groups[0].Entries = append(hf.Groups[0].Entries, entry)
}

// addEntryToGroup adds an entry to a specified group (creates group if it doesn't exist)
func addEntryToGroup(hf *domain.HostsFile, entry domain.Entry, groupName string) error {
	// Find or create the group
	for i := range hf.Groups {
		if hf.Groups[i].Name == groupName {
			hf.Groups[i].Entries = append(hf.Groups[i].Entries, entry)
			return nil
		}
	}

	// Group doesn't exist, create it
	newGroup := domain.Group{
		Name:        groupName,
		Description: "",
		Entries:     []domain.Entry{entry},
		Groups:      []domain.Group{},
	}
	hf.Groups = append(hf.Groups, newGroup)
	return nil
}
